### ISSUE TITLE
"Styled(Component) > Unknown" always top of react-addons-perf wasted list

### ISSUE THREAD

**Issue by ggascoigne** [styled-components/styled-components#875]
First let me say that I don't think that this is necessarily a bug with styled components, I honestly suspect my own code. Secondly, much of this is pretty new to me so please forgive what I fear might be a bit of a noob question.

That said, I'm using react-addons-perf to try and understand why parts of my app appear to render slowly and I'm consistently seeing that "Styled(Component) > Unknown" is the number one function in the wasted time list, regularly being responsible for over 50% of the elapsed time and having a higher render count that pretty much all of the other "wasted" operations put together.

And I'm trying to understand exactly what this is telling me.

We don't use styled-components directly, we pull it in via flexbox-react, and we use Flexbox all over the place.

Given that the Flexbox is a styled container class, can you suggest anything that I can do to get a better handle on where the time is being spent?

**Comment by k15a** [comment]
Do you have any example code which you know that it has bad performance and you can share? I think this is a problem with react-flexbox.

**Comment by philpl** [comment]
You'll need to make sure that the structural components and containers above don't trigger a rerender when their data doesn't actually change. On our end we can't do much, since we're quite low in the tree. Adding a pure check with referential equality would just introduce unexpected behaviour for beginners :)

**Comment by probablyup** [comment]
@philipl it would be cool if there was a way to somehow turn pure mode on.

**Comment by philpl** [comment]
@probablyup our "philosophy", if you can call it that, is to not provide any options/switches/modes, or similar. This might change in the future, if we find a non-compromise way to do it nicely, but we'd like to get to a single good experience, with a "linear progression" of use-cases and techniques.

In other words, it's unlikely that we'll introduce a mode for this, or a method, or anything similar.

On the other hand in this case there's a very important reason why it's irrelevant whether there's a pure mode or not: Styled Components is all about presentational components. This means, unlike structural or container components, they'll always tend to be at the bottom of a component tree.

Whatever architecture you're applying, as long as you're already adhering to some "flavour" of React best practices, we can always assume that there'll always be a container component or structural component above, that is actually determining the external state i.e. props for a presentational component / styled component –– *sounds far afield but...*

When you're passing in derived data into a presentational component, there'll always be a higher component that can prevent an unnecessary re-render in a more elegant, efficient, and intuitive way than what's necessary or possible in a presentational component.

This doesn't exclusively apply to styled components of course. Therefore I don't think that a pure mode for SC is even desirable in the first place.

On top of this, I'd also like to add –– this is of course more of an assumption –– that performance problems in React often don't stem from the lowest components. Far more often I've seen the truly expensive operations and updates happening earlier, higher up the tree.

----

On that note, do you guys think this needs to be added to our FAQ? It might be a useful hint to more.

**Comment by probablyup** [comment]
@philpl Hmm I sort of get what you mean. If you're a library author though offering styled components though, I don't think that necessarily holds true. For example, at Zocdoc we maintain a component SDK that has more than 20 styled components within. We routinely plop these into our frontend and add styled-component wrappers or simple style extension to these.

In that sort of situation, being able to toggle on "pure" mode would be beneficial I think. I do understand where you're coming from though. The only counter argument I'd make there is if people use this library, they're probably using it _a lot_. As in hundreds to thousands of component instances on a single page. That pure optimization could make a very big difference to the render performance of such a page.

**Comment by philpl** [comment]
@probablyup counterintuitively all still holds true even in this case. If your base components accept specific data, there'll be a structural components that you can use for this. If not, there'll only be a Styled component so it won't be necessary at all (see above post).

There are a truck load of arguments why we don't need this and another one is, that it's an explicit effect, that should only be applied when it makes sense from a state and data standpoint, not a styling one. There's some good patterns to create a HOC for this. We simply don't need this option.

Also, if you're rerendering hundreds of components at the same time. All styling libs will be too slow. We can't make any assumption about input and output, so we have to recompute styles. Simply put, if any changes unnecessarily reach SC, there's always a higher component that is at fault. You can also think about this from a different perspective and say, that every time you go a node up in your component tree, you target more components. Thus preventing updates higher up is always better.

It's irrelevant and which level your styled components come in. They'll simply always be the lowest components. If you're wrapping it doesn't matter either, you can either turn it into a different pattern if needed, or it's inexpensive anyway.

So to sum up: there's different solutions (eg a pure() hoc), it doesn't make sense to control data and updates in a styling lib, we're at the lowest points of the tree, container components are a better place to put this, and if you're rerendering often sth has already gone wrong.

I hope this is a sufficient reply. mind you that writing them does take out a lot of time for other SC work, so I hope that it's enough material now ;)

**Comment by mxstbr** [comment]
> That pure optimization could make a very big difference to the render performance of such a page.

Sidenote: **It'll likely make your page slower to do that many shallow equality checks.** Most of the time those shallow equality checks are more expensive than just rendering the component again, that's why `PureComponent` isn't the default in React.

So **adding a shallow equality check to styled-components would very likely make it slower than before**. The best way to use `shouldComponentUpdate` is to bail out as high up in your component tree as you can, to avoid re-rendering as much of the app as possible with a single check.

**Comment by probablyup** [comment]
@philpl thanks for taking the time to discourse!

@mxstbr it probably wouldn't be slower. Keep in mind that pure components abort updates for their entire subtree if a prop change isn't detected. If there are changes all the way down the tree and it has to do a pure check on every level, sure that would take slightly more time.

It's absolutely true that the first render will always be slow-ish with a great many components, but once a tree has been formed and React is just diffing against it, using pure mixins/components and immutable data structure / philosophies helps React save a lot of effort.

**Comment by mxstbr** [comment]
I'm pretty sure that's offset by the fact that it's slow in the critical path where it's rendering everything, otherwise React would've made `PureComponent` the default behaviour rather than an opt-in enhancement. :wink:

**Comment by probablyup** [comment]
Not really. The React team is pretty compatibility-minded and it's a different prop behavior. They have encouraged the use of SFC where possible though, which gets the pure optimization for free.

**Comment by mxstbr** [comment]
> They have encouraged the use of SFC where possible though, which gets the pure optimization for free.

That's actually incorrect, **function components are not pure**. Quite on the contrary, they act like any class component and always re-render. (see [this demo](https://www.webpackbin.com/bins/-Km2NG4tkkwK5hXYaR5v) and look at the console)

Why? Because, like I said, doing a shallow equality check at every component would be slower than just rendering it to the virtual DOM. Since you don't seem to believe me I dug up [this tweet](https://twitter.com/dan_abramov/status/820668074223353858) just for you: 

> PSA: React.PureComponent can make your app slower if you use it everywhere.
> - Dan Abramov

And [this one](https://twitter.com/dan_abramov/status/759383530120110080) just for good measure:

> Don‘t use PureComponent everywhere. Measure.
> - Dan Abramov

Do you believe me now that it might not be the best idea to make every styled component be a pure component?

**Comment by probablyup** [comment]
@mxstbr I must have confused the verbiage around SFCs... I remember when they first came out there was talk about them being pure by default but must be mistaken or not remembering correctly.

> Do you believe me now that it might not be the best idea to make every styled component be a pure component?

Perhaps not, but being able to turn that on via `.withConfig()` or something so the developer has control isn't an absurd ask.

**Comment by philpl** [comment]
@probablyup they're functionally pure, which is different. See my posts above for several reasons why it's neither necessary, and why we won't add it

**Comment by ggascoigne** [comment]
I just wanted to chime in and thank you all for this discussion, it's been really educational :)  After some digging, I'm fairly sure that my problem was caused by an overzealous use of SFC way too high up my component tree, some surgical use of `shouldComponentUpdate()` is making a huge improvement.

**Comment by mxstbr** [comment]
> Perhaps not, but being able to turn that on via .withConfig() or something so the developer has control isn't an absurd ask.

It's not absurd, but also doesn't really make sense considering it very likely makes your app slower :wink: (on top of `styled-components` not having options)

```JS
styled.set({
  makeMyAppSlow: true,
});
```

**Comment by mxstbr** [comment]
> some surgical use of shouldComponentUpdate() is making a huge improvement.

That's great to hear @ggascoigne!

**Comment by probablyup** [comment]
I meant on a per-component basis like `styled.pure.div` etc, but thanks for the snarkiness.
